feat: summary status bar and improve process handling

justfalter opened the pull request:
This PR started with a desire to add a summary status bar, but I quickly found some improvements were required to command handling to make that happen.

- Improves visibility into the state of the extension by adding a `Run On Save Status` status bar. Text of  `running Sync: 3 Async: 2` indicates that the extension is currently running 3 synchronous and 2 async commands. Text of `idle Sync: 0 Async: 0` indicates that nothing is currently running.
    - Clicking the status bar reveals the Run On Save output.
 - Removes vestigial status bar code that stopped working a long time ago.
- Improves synchronous command handling:
  - Sync commands are processed one at a time in FIFO order. Prior to this change, saving 20 files at the same time (like with a search/replace) would lead to 20 separate sync commands being run simultaneously. With this change, the 20 commands would be executed one at a time.
- When disabled, terminates all running commands. This provides a much-needed means to terminate commands when things have gone awry.
  - The signal used to kill the processes is controlled via command's `killSignal` property, defaulting to `SIGTERM`.

## Description of Changes
Summarize changes this PR introduces.

## Testing

### Poke around with the status bar.

Have run on save config, verify that the status bar displays expected information.

### Verify disabling extension terminates running commands.
The following command will spawn sync and async commands, each starting a "sleep":

```jsonc
{
  "emeraldwalk.runonsave": {
    "commands": [
      {
        "match": ".*",
        "cmd": "sleep 99999"
      },
      {
        "match": ".*",
        "cmd": "sleep 88888",
        "isAsync": true
      }
    ]
  }
}
```

1. Save several files
2. Observe that the number of sync and async commands have counted upward.
3. run `ps aux| grep sleep` and observe the accumulation of a `sleep 99999` sync command and several `sleep 88888` async commands.
4. Run the `Run On Save: Disable` command.
5. Observe that the processes have been terminated and the queue cleared.

## Checklist
- [x] Did you update any relevant docs / samples?
- [x] Did you include details on how to test your changes?

justfalter committed:
feat: summary status bar and improve process handling

This PR started with a desire to add a summary status bar, but I quickly found
some improvements were required to command handling to make that happen.

- Improves visibility into the state of the extension by adding a
  `Run On Save Status` status bar. Text of  `running Sync: 3 Async: 2`
  indicates that the extension is currently running 3 synchronous and 2 async
  commands. Text of `idle Sync: 0 Async: 0` indicates that nothing is currently
  running.
    - Clicking the status bar reveals the Run On Save output.
- Removes vestigial status bar code that stopped working a long time ago.
- Improves synchronous command handling:
  - Sync commands are processed one at a time in FIFO order. Prior to
    this change, saving 20 files at the same time (like with a search/replace)
    would lead to 20 separate sync commands being run
    simultaneously. With this change, the 20 commands would be executed
    one at a time.
- When disabled, terminates all running commands. This provides a
  much-needed means to terminate commands when things have gone awry.
  - The signal used to kill the processes is controlled via
    command's `killSignal` property, defaulting to `SIGTERM`.

bmingles commented on README.md:
```diff
@@ -23,7 +41,8 @@ Add "emeraldwalk.runonsave" configuration to user or workspace settings.
   - `notMatch` - a regex for matching which files **NOT** to run. Files that match this pattern take precedence over ones that match the `match` option (see [Notes on RegEx Options](#notes-on-regex-options)).
   - `cmd` - command to run. Can include parameters that will be replaced at runtime (see Placeholder Tokens section below).
-  - `isAsync` (optional) - defaults to false. If true, next command will be run before this one finishes.
+  - `isAsync` (optional) - defaults to false, meaning that the command wait in a FIFO queue while other synchronous commands are being processed one at a time. If true, next command will be run before this one finishes.
```
```suggestion
  - `isAsync` (optional) - defaults to false, meaning that the command waits in a FIFO queue while other synchronous commands are being processed one at a time. If true, next command will be run before this one finishes.
```

bmingles commented on src/extension.ts:
```diff
@@ -162,12 +180,48 @@ class RunOnSaveExtension {
+    if (notifyWaitingRunner) {
+      this._notifyWaitingSyncCommandRunner = undefined;
+      notifyWaitingRunner();
+    }
```
I think this can be simplified:
```suggestion
    this._notifyWaitingSyncCommandRunner?.()
    this._notifyWaitingSyncCommandRunner = undefined;
```

bmingles reviewed (changes requested):
@justfalter I've read through the diff and left a few comments.

I still need to actually test and make sure I fully understand the new queueing mechanism.

I also specifically want to test in Windows to make sure there are no surprises with the kill signal configurations since not all are supported by Windows.

Thanks again for the PR. I'm glad to have real users fixing real problems they are experiencing.

bmingles commented on src/extension.ts:
```diff
@@ -208,21 +263,44 @@ class RunOnSaveExtension {
+    const atLeastOneCommandIsRunning = unfinishedSyncCount !== 0 || asyncCount !== 0;
+
+    if (!this.isEnabled()) {
+      // If we are disabled but have unfinished commands, then we are "draining".
```
What's the scenario that can result in "draining"? Since commands are killed when extension is disabled, is there any meaningful time that we would see this status?

bmingles commented on src/extension.ts:
```diff
@@ -208,21 +263,44 @@ class RunOnSaveExtension {
+    }
+
+    this._sbStatus.text = `${state} Sync: ${unfinishedSyncCount} Async: ${asyncCount}`;
+    this._sbStatus.tooltip = `Run On Save: State: ${state}, Unfinished synchronous command count: ${unfinishedSyncCount}, Active asynchronous command count: ${asyncCount}`;
```
I like the status summary idea, but a few things I'd like to tweak:

1. The status bar summary could be a little more compact. It also currently has a layout shift. This can be fixed by using icons instead of so much text.
2. The `sync` vs `async` verbiage is an unfortunate artifact of an early version of the extension. In practice it's more like `sequential` vs `parallel`. It's a bit too late to change the config setting (`isAsync`), but I'd like to change it in labels and such going forward to be more accurate

- Running 
   ![running-status](https://github.com/user-attachments/assets/85ab7803-7d21-4b18-a39a-2f86fde4d141)
- Disabled
   <img width="50" height="28" alt="image" src="https://github.com/user-attachments/assets/de10740e-d675-4d34-b4e5-517462ce27b4" />
- Idle
   <img width="86" height="28" alt="image" src="https://github.com/user-attachments/assets/3fb21c5c-aa0e-4b6e-a972-48c2cd57b5f7" />
- Draining (if we need this) 
   <img width="84" height="27" alt="image" src="https://github.com/user-attachments/assets/7ffbaaa5-fe69-4651-ba6f-ac6b5bb26da4" />


3. I think we can simplify the tooltip a bit as well

```suggestion
    let state: 'Draining' | 'Disabled' | 'Idle' | 'Running';

    const atLeastOneCommandIsRunning =
      unfinishedSyncCount !== 0 || asyncCount !== 0;

    if (!this.isEnabled()) {
      // If we are disabled but have unfinished commands, then we are "draining".
      state = atLeastOneCommandIsRunning ? 'Draining' : 'Disabled';
    } else {
      // If we are enabled but have no unfinished commands, then we are "idle".
      state = atLeastOneCommandIsRunning ? 'Running' : 'Idle';
    }

    const statsShort = `S:${unfinishedSyncCount},P:${asyncCount}`;
    const statsLong = `Sequential: ${unfinishedSyncCount} Parallel: ${asyncCount}`;

    const text = {
      Draining: `$(clock) ${statsShort}`,
      Disabled: '$(save) $(circle-slash)',
      Idle: `$(save) ${statsShort}`,
      Running: `$(loading~spin) ${statsShort}`,
    } as const;

    this._sbStatus.text = text[state];
    this._sbStatus.tooltip = `Run On Save: ${state}${
      state === 'Disabled' ? '' : `, Remaining commands: ${statsLong}`
    }`;
```

bmingles reviewed (changes requested):
@justfalter First of all, thank you for your contribution. I appreciate the help in improving the extension.

### Inline Comments
I've done some testing and reviewed the changes a bit closer. I've left a handful of inline comments in the code.

### Command Management Changes
Regarding the command management changes, there are 2 primary concerns I have related to changes in existing functionality.

1. Parallel commands (`isAsync:true`) that are configured after sequential ones (`isAsync:false`) used to wait until the preceeding sequential command completed. It seems that the parallel commands execute immediately now. They should wait for prior sequential commands to finish. For reference, I used this configuration to test:

```jsonc
"emeraldwalk.runonsave": {
      // Messages to show before & after all commands
      "message": "*** All Start ***",
      "messageAfter": "*** All Complete ***",
      // Show elappsed time for all commands
      "showElapsed": true,
      "commands": [
        {
          // This cmd should block all following commands (sequential and parallel ones)
          "match": ".*",
          "cmd": "sleep 3",
          "message": "1. sleep 3"
        },
        {
          // This cmd should wait until the first finishes
          "match": ".*",
          "cmd": "sleep 10",
          "message": "2. sleep 10",
          "isAsync": true
        },
        {
          "match": ".*",
          "cmd": "sleep 4",
          "message": "3. sleep 4",
          "isAsync": true
        }
}
```

2. I think your observation that saving multiple files can cause many exec commands to fire at once is a good one and something I did not account for. My concern with the current changes is that it may impact existing users that are expecting the pre-existing behavior. For example, a user may actually want 20 exec commands to fire off in parallel across multiple files. Changing them to run sequentially across files could introduce unexpected lag in certain cases. I think it's a good fix, but it should be opt-in via a setting of some sort.

This does introduce some additional complexity since the command queue triggered by each file save will need to honor the rule mentioned in bullet 1 but also be able to order sequential commands across files if a user opts in to this feature.

Thanks again for the PR. Let me know if you need any clarification on my comments.

justfalter committed:
move RunCommandConfig to model.ts

justfalter committed:
make enable/disable async

justfalter committed:
Apply suggestions from code review

Co-authored-by: Brian Ingles <github@emeraldwalk.com>

justfalter committed:
restore processing of async commands in config order

- Match prior behavior of only launching async commands when encountered
  in config order. We do not wait for an async command to complete
  before starting the next command.

justfalter commented:
Life has gotten bonkers and I haven't been able to throw time at this, but I'm taking a look again, now.
